app/services/parent_profile_associator.rb

```diff
@@ -6,26 +6,74 @@ class << self
     def run!
       Profile.transaction do
         Profile.where.not(account: nil).find_each do |profile|
-          profile.update!(parent_profile: find_parent(profile))
+          parent = find_parent(profile)
+          next if profile_already_in_account!(profile, parent)
+
+          profile.update!(parent_profile: parent, benchmark: parent.benchmark)
         end
       end
     end
 
     private
 
+    def profile_already_in_account!(profile, parent)
+      duplicate_profiles = profile.account.profiles.where(
+        ref_id: profile.ref_id,
+        benchmark_id: parent.benchmark.id
+      ).count
+      return false unless duplicate_profiles >= 1
+
+      profile.delete_all_test_results = true
+      profile.destroy
+    end
+
     def find_parent(profile)
-      parents = Profile.where(ref_id: profile.ref_id,
-                              benchmark_id: profile.benchmark_id,
-                              account_id: nil)
-      raise not_found(profile) unless parents.one?
+      logger.info("Finding parents for profile #{profile.id}")
+      parents = find_in_same_benchmark(profile)
+      parents = find_in_other_benchmarks(profile) if parents.blank?
+
+      raise not_found(profile) unless parents.any?
+
+      parent = find_most_recent_parent(parents)
+      logger.info("Found parent #{parent.id} for profile #{profile.id}")
+      parent
+    end
+
+    def find_in_same_benchmark(profile)
+      Profile.where(ref_id: profile.ref_id,
+                    benchmark_id: profile.benchmark_id,
+                    account_id: nil)
+    end
+
+    # Profiles in the phony-benchmark will have to be assigned to
+    # a real benchmark if it can be found.
+    def find_in_other_benchmarks(profile)
+      parent_matching_description = Profile.where(
+        name: profile.name,
+        ref_id: profile.ref_id,
+        description: profile.description,
+        account_id: nil
+      )
+      return parent_matching_description if parent_matching_description.present?
```

Comment:
Regarding @akofink comment on https://github.com/RedHatInsights/compliance-backend/pull/357/files#r399620519 - this fixes it IMO - if we can find the exact parent based on name, ref_id and description, good, we do it. Otherwise, we just match for the next better option, which is just name/ref_id. 

Initially I did this with a `or` query, but the AR result doesn't really fit properly, so I'm resolving each AR relation in order (it's not terribly slow on the DB dump we have). Notice `parent_matching_description` is `[]` if there are no results, so I can't do `a || b` either.  

```
[17] pry(main)> Profile.where(name: profile['name'], ref_id: profile['ref_id'], description: profile['description'], account_id: nil).or(Profile.where(name: profile['name'], ref_id: profile['ref_id'], account_id: nil)).to_sql
=> "SELECT \"profiles\".* FROM \"profiles\" WHERE \"profiles\".\"name\" = 'OSPP - Protection Profile for General Purpose Operating Systems v. 4.2' AND \"profiles\".\"ref_id\" = 'xccdf_org.ssgproject.content_profile_ospp42' AND \"profiles\".\"account_id\" IS NULL"

```

Reply:
My only concern is that we really only want this for profiles with a description that matches the benchmark description (i.e. `This guide presents a catalog of security-relevant\nconfiguration settings...`), but from my testing, those are the only profiles that won't be found in another benchmark anyway.